lib/git/base.rb

```diff
@@ -198,46 +216,53 @@ def lib
       @lib ||= Git::Lib.new(self, @logger)
     end
 
-    # will run a grep for 'string' on the HEAD of the git repository
-    #
-    # to be more surgical in your grep, you can call grep() off a specific
-    # git object.  for example:
-    #
-    #  @git.object("v2.3").grep('TODO')
+    # Run a grep for 'string' on the HEAD of the git repository
     #
-    # in any case, it returns a hash of arrays of the type:
-    #  hsh[tree-ish] = [[line_no, match], [line_no, match2]]
-    #  hsh[tree-ish] = [[line_no, match], [line_no, match2]]
+    # @example Limit grep's scope by calling grep() from a specific object:
+    #   git.object("v2.3").grep('TODO')
     #
-    # so you might use it like this:
-    #
-    #   @git.grep("TODO").each do |sha, arr|
+    # @example Using grep results:
+    #   git.grep("TODO").each do |sha, arr|
     #     puts "in blob #{sha}:"
-    #     arr.each do |match|
-    #       puts "\t line #{match[0]}: '#{match[1]}'"
+    #     arr.each do |line_no, match_string|
+    #       puts "\t line #{line_no}: '#{match_string}'"
     #     end
     #   end
+    #
+    # @return [Hash<String, Array>] a hash of arrays
+    #   ```Ruby
+    #   {
+    #      'tree-ish1' => [[line_no1, match_string1], ...],
+    #      'tree-ish2' => [[line_no1, match_string1], ...],
+    #      ...
+    #   }
+    #   ```
+    #
     def grep(string, path_limiter = nil, opts = {})
       self.object('HEAD').grep(string, path_limiter, opts)
     end
 
     # updates the repository index using the working directory content
     #
-    #    @git.add('path/to/file')
-    #    @git.add(['path/to/file1','path/to/file2'])
-    #    @git.add(:all => true)
+    # @example
+    #   git.add
+    #   git.add('path/to/file')
+    #   git.add(['path/to/file1','path/to/file2'])
+    #   git.add(:all => true)
     #
     # options:
     #   :all => true
     #
     # @param [String,Array] paths files paths to be added (optional, default='.')
     # @param [Hash] options
-    def add(*args)
-      if args[0].instance_of?(String) || args[0].instance_of?(Array)
-        self.lib.add(args[0],args[1]||{})
-      else
-        self.lib.add('.', args[0]||{})
-      end
+    # @option options [boolean] :all
+    #   Update the index not only where the working tree has a file matching
+    #   <pathspec> but also where the index already has an entry.
+    #   See [the --all option to git-add](https://git-scm.com/docs/git-add#Documentation/git-add.txt--A)
+    #   for more details.
+    #
+    def add(paths = '.', **options)
```

Comment:
Change the signature from `*args` so the individual options can be documented.  I am confident that this is equivalent because the unit tests still pass.